server/src/test/java/io/spine/examples/shareaware/server/e2e/SharePurchaseTest.java

```diff
@@ -0,0 +1,93 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.shareaware.server.e2e;
+
+import io.grpc.ManagedChannel;
+import io.spine.examples.shareaware.investment.InvestmentView;
+import io.spine.examples.shareaware.server.e2e.given.E2EUser;
+import io.spine.examples.shareaware.server.e2e.given.SubscriptionOutcome;
+import io.spine.examples.shareaware.server.e2e.given.WithServer;
+import io.spine.examples.shareaware.share.Share;
+import io.spine.examples.shareaware.wallet.WalletBalance;
+import io.spine.examples.shareaware.wallet.rejection.Rejections.InsufficientFunds;
+import io.spine.server.tuple.EitherOf2;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static com.google.common.truth.extensions.proto.ProtoTruth.assertThat;
+import static io.spine.examples.shareaware.given.GivenMoney.usd;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.balanceAfterTeslaPurchase;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.insufficientFundsWith;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.investmentAfterTeslaPurchase;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.pickTesla;
+import static io.spine.examples.shareaware.server.e2e.given.SharePurchaseTestEnv.zeroWalletBalance;
+
+/**
+ * End-to-end test that describes such a scenario:
+ * <ol>
+ *     <li>The user attempts to purchase shares with no money in the wallet.</li>
+ *     <li>The user replenishes his wallet for 500 dollars.</li>
+ *     <li>The user successfully purchases one 'Tesla' share.</li>
+ *     <li>The user withdraws all his money from the wallet.</li>
+ * </ol>
+ */
+final class SharePurchaseTest extends WithServer {
+
+    @Test
+    @DisplayName("User should purchase one tesla share and withdraw all the money after this")
+    void test() {
+        ManagedChannel channel = openChannel();
+        E2EUser user = new E2EUser(channel);
+
+        List<Share> shares = user.waitsForSharesToUpdate();
+        Share tesla = pickTesla(shares);
+
+        EitherOf2<WalletBalance, InsufficientFunds> failedPurchase = user.purchase(tesla, 1);
```

Comment:
This operation also seems fine. We send a command and claim we may either get a balance in return, or a rejection.

But inside of `purchase`, we query the balance instead of subscribing to it. Can we migrate this to a subscription?